### DISCUSSION TITLE
Protobuf service types

### DISCUSSION THREAD

**Discussion by stychu** [connectrpc/connect-es#796] [Q&A] [answered]
Hi there,

I'm intrigued by `connect-es` and want to try it asap. But I have one question.

Let's say I have a React application with Express (tRPC + zod) acting as a middleware server (proxy) to the external API (source of truth/data). 

Flow look like this:
`Client Req -> tRPC handler -> validate request (zod) -> makes fetch call to the external API -> parse response from the API (zod) -> Client Response`

Having zod in the mix is very helpful and useful because I have an access not only to the TypeScript types of the `API Response` but auto validation of it thus strong type inference which I can also use on the Client.

I am very new to Protocol Buffers and from what I read in the `connect-es` [docs](https://connectrpc.com/docs/node/implementing-services)  (which are pretty great btw! Good Job <3)

> Connect handles HTTP routes and most plumbing for you, but implementing the actual business logic is still up to you.

 that when Registering a Service  I would need to define Response/Request types of the message and that business logic is only  a fetch to the API and passing down the response leads me to think that I would have duplicate `schemas` for my external API. 

1. zod API Request/Response 
2. Service Request/Response (that I would need to write by hand again)

So question is:
1. Am I forced to define a Service proto type file write again types of API and basically ditch zod in this case?
2. Can I reuse types from zod schemas for Service request/response and not duplicate typings in Service proto?

**Comment by srikrsna-buf** [comment, marked as answer]
Hey! To use connect-es you have to define the types in protobuf. You will get similar type inference, if not better, on the client side when using connect-es. You can still use zod for validations like email etc... and protobuf will ensure the basic types (string/number) are as expected.

We are working on https://github.com/bufbuild/protovalidate, which aims to bring validation to protobuf messages. It already has support for several languages with TS being actively worked on.

**Comment by stychu** [comment]
So lets say I have Prisma in my application stack .

```ts
// Prisma model
model Post {
  id    String @id @default(uuid())
  title String
  text  String
}
```

```ts
// prisma call implementation

async ({ id }) => {
      const post = await prisma.post.findUnique({
        where: { id },
      });

      return post;
    })
```

This implementation call is automatically type assisted thanks to the prisma model so return of such call is known as 
```ts
{
	id: 'someid',
	title: 'sometitle',
	text: 'sometext',
}
```


But If I would like to use prisma  with `connect-es` I would need to again declare `Response` type in the protobuf Service file despite having it already at a hand provided by the prisma model??

```ts
message Request {
  string id = 1;
}
message Response {
  string id = 1;
  string title = 2;
  string text = 3;
}
service PostService {
  rpc byId(Request) returns (Response) {}
}
```
```ts
...
export default (router: ConnectRouter) =>
  router.service(PostService, {
    async byId(req: Request, context: HandlerContext) {
	const post = await prisma.post.findUnique({
	        where: { id },
	      });

      return new Response(post);
    }
  });
```


Or can I reuse the Types provied by the prisma model and omit definition of the Request/Response in the proto Service file ?

**Reply by nicksnyder** [reply]
Yes you need to declare your types in Protobuf.
